internal/rukpak/source/unpacker.go

```diff
@@ -0,0 +1,120 @@
+type Unpacker interface {
+	Unpack(context.Context, *rukpakapi.BundleSource, *clusterextension.ClusterExtension) (*Result, error)
+}
+
+// Result conveys progress information about unpacking bundle content.
+type Result struct {
+	// Bundle contains the full filesystem of a bundle's root directory.
+	Bundle fs.FS
+
+	// ResolvedSource is a reproducible view of a Bundle's Source.
+	// When possible, source implementations should return a ResolvedSource
+	// that pins the Source such that future fetches of the bundle content can
+	// be guaranteed to fetch the exact same bundle content as the original
+	// unpack.
+	//
+	// For example, resolved image sources should reference a container image
+	// digest rather than an image tag, and git sources should reference a
+	// commit hash rather than a branch or tag.
+	ResolvedSource *rukpakapi.BundleSource
+
+	// State is the current state of unpacking the bundle content.
+	State State
+
+	// Message is contextual information about the progress of unpacking the
+	// bundle content.
+	Message string
+}
+
+type State string
+
+const (
+	// StatePending conveys that a request for unpacking a bundle has been
+	// acknowledged, but not yet started.
+	StatePending State = "Pending"
+
+	// StateUnpacking conveys that the source is currently unpacking a bundle.
+	// This state should be used when the bundle contents are being downloaded
+	// and processed.
+	StateUnpacking State = "Unpacking"
+
+	// StateUnpacked conveys that the bundle has been successfully unpacked.
+	StateUnpacked State = "Unpacked"
+)
+
+type unpacker struct {
+	sources map[rukpakapi.SourceType]Unpacker
+}
+
+// NewUnpacker returns a new composite Source that unpacks bundles using the source
+// mapping provided by the configured sources.
+func NewUnpacker(sources map[rukpakapi.SourceType]Unpacker) Unpacker {
+	return &unpacker{sources: sources}
+}
+
+func (s *unpacker) Unpack(ctx context.Context, bs *rukpakapi.BundleSource, ce *clusterextension.ClusterExtension) (*Result, error) {
+	source, ok := s.sources[bs.Type]
+	if !ok {
+		return nil, fmt.Errorf("cluster extension %q, source type %q not supported", ce.Name, bs.Type)
+	}
+	return source.Unpack(ctx, bs, ce)
+}
+
+// NewDefaultUnpacker returns a new composite Source that unpacks bundles using
+// a default source mapping with built-in implementations of all of the supported
+// source types.
+//
+// TODO: refactor NewDefaultUnpacker due to growing parameter list
+func NewDefaultUnpacker(mgr manager.Manager, namespace, unpackImage string, rootCAs *x509.CertPool) (Unpacker, error) {
+	cfg := mgr.GetConfig()
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	httpTransport := http.DefaultTransport.(*http.Transport).Clone()
+	if httpTransport.TLSClientConfig == nil {
+		httpTransport.TLSClientConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+		}
+	}
+	httpTransport.TLSClientConfig.RootCAs = rootCAs
```

Comment:
I think we don't need the TLS client config or the rootCAs. IIRC it was used only in http source.

Reply:
Hmmm... even in the original code, it doesn't appear used. It's a clone (copy?) and isn't even referenced after this. Odd...